Feat/phase system

OrcaOfTime opened the pull request:
Previously the phase logic was tightly coupled in to the Task Manager. Making it difficult to add any more phases if required.

I have extracted that logic and data into 'ITaskPad' classes. The interface gives us the generic logic allow us to add as many phases as we like whilst allowing for custom logic AND UIs within each phase e.g. the format of tasks listed on the Taskpad bein all available tasks in phase 1 then only showing one task at a time in phase 2. 

I have also changed the naming convention of the phases to instead be a short descriptive name rather than an arbitrary number in case later down the line we want to add phases in between or even before the current phase order, which could cause confusion in the back end. 

Testing - 
Game should flow as normal transitioning from daily task phase to weapon phase

OrcaOfTime committed:
feat(code): Phase SO created

OrcaOfTime committed:
feat(code): Added Task list and some temp note comments

OrcaOfTime committed:
feat(code): Transferred task complete noises to phase SOs

I am a genius

OrcaOfTime committed:
feat(code): Small phase and unity scene change

I'm not sure what exactly this change is, I was just prompted to save before closing and got scared I missed something

OrcaOfTime committed:
feat(code): Transferring taskpad logic into separate script

Starting an attempt to generalise the UI updating logic of the Task Pad whilst maintaining the freedom to make a unique taskpad UI for each phase.
No all the relevant data can be stored in a scriptable object so I have opted for placing each phase UI canvas into a list on a Task Pad script, this should now just switch the Active canvas to whichever phase we have reached.
Any common elements like task list remain in a scriptable object for the time being

OrcaOfTime committed:
feat(code): Order task file structure

Includes intro to taskpad script but that was included in the restructure so didn't want to leave it out of the commit

OrcaOfTime committed:
feat(code): Transfering head colour change logic to phase script

OrcaOfTime committed:
feat(code): Reformatting Phase data file structure

Another quick rename and restructure of the phase and phaseData scripts

OrcaOfTime committed:
feat(code): transferring functionality to taskPad scripts

Removing the phase functionality out to the now more generic TaskPad classes.
This should allow us the ability to have more than one phase but still able to make custom UIs for them

OrcaOfTime committed:
feat(git): Added random asset file changes to git ignore

These changes are driving me insane. Ignore their relentless wails

OrcaOfTime committed:
feat(code): Rename phase 2 flag

Committed to distinguishing phases by name instead of arbitrary number

OrcaOfTime committed:
feat(code): fix TaskPad UI quirks

Re-instated normal behaviour on taskPad UIs

OrcaOfTime committed:
feat(code): more reset logic when pacifist ending reached

OrcaOfTime committed:
Merge branch 'develop' into feat/phase-system

OrcaOfTime committed:
feat(code): Timed event changes included in weaponTaskPad

OrcaOfTime committed:
feat(code): Moved phasesData into ITaskPad object instead of SO

The phase data was pointless in the end with the final structure I created

PiersSK commented on .gitignore:
lol thank you

OrcaOfTime committed:
feat(code): I FORGOT TO SAVE UNITY

Commander lines removed in this commit

PiersSK reviewed (changes requested):
Some initial spots

- At start of weapons phase, weapons process overview on taskpad reads 0/0 instead of 0/6 tasks complete
- CM cabin door opens after weapon warmup and not after calibration as expected
    - This allows players to complete the unlock weapon task earlier than they're supposed to
- Fire button throws error, seems to be linked to the commander voice lines you called out in dev

Generally it looks great though, but a couple of bits feel a little hardcodey still. Having isWeaponPhase is still maybe too specific as not every level will have a weapons phase necessarily. Maybe we change it to check for a certain phase index, or give things a phase object to explicitely comare to the current??

OrcaOfTime committed:
feat(code): correct weapon task count

No tasks exist on awake silly, and with the data restructure I did you can just fucking GRAB the total task counts when you want

OrcaOfTime committed:
feat(code): rename taskpad files to Phase

OrcaOfTime committed:
feat(code): Fire weapon exception fixed

Turns out I don't know how to avoid index exceptions in lists. Fucking idiot

OrcaOfTime committed:
feat(code): Removed isWeaponPhase check (interim logic)

The logic added here checks for an explicit class type 'WeaponTaskPhase' which is fine for the task scripts as they are only going to be relevant for  that one task.
However there is some logic in managers that operate on a single level basis. For example the time controller setting a timer.
I will add a ticket to generalise this a bit more so we are looking for a 'type' of phase instead of a specific phase in a specific level

OrcaOfTime committed:
feat(code): Mandatory forget to save on the scene editor

PiersSK committed:
fix(code): correctly orders daily tasks in editor

PiersSK committed:
fix(code): adds new commander mute on game end
